### PR TITLE
Implement new PoseBroadcaster controller

### PR DESCRIPTION
**Pull request by RobertWilbrandt**
This PR implements a new broadcasting controller based on the `PoseSensor` semantic component introduced in https://github.com/ros-controls/ros2_control/pull/1775.

As described in the `PoseSensor` PR, my primary use-case for this is the publishing of tcp poses from robot arms. This can be more accurate than calculating forward kinematics, as robot calibrations etc. can be considered by the robot. Settings on the robot side concerning tool offsets etc. can also be included by this (at least for UR and kuka RSI). The ROS 1 `ur_robot_driver` used to publish this transform directly from within the hardware interface, but in ROS 2 there is not yet a standard way for providing this information. A PR for the UR ROS 2 driver already exists at https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/pull/1108.

This controller publishes poses directly as `geometry_msgs::msg::PoseStamped`s and as tf transforms (can be disabled).

### CODE REVIEW

File: pose_broadcaster/pose_broadcaster.xml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<library path="pose_broadcaster">
+  <class name="pose_broadcaster/PoseBroadcaster"
+         type="pose_broadcaster::PoseBroadcaster"
+         base_class_type="controller_interface::ControllerInterface">
+    <description>
+      This controller publishes a cartesian state as a geometry_msgs/PoseStamped message.
```
Comment (fmauch) [inline pose_broadcaster/pose_broadcaster.xml:6]: **Very** nitpicky, but AFAIK Cartesian is spelled with a capital C.
```suggestion
      This controller publishes a Cartesian state as a geometry_msgs/PoseStamped message.
```

File: pose_broadcaster/src/pose_broadcaster.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+  if (realtime_tf_publisher_ && realtime_tf_publisher_->trylock())
+  {
+    bool do_publish = false;
+    // rlcpp::Time comparisons throw if clock types are not the same
+    if (tf_last_publish_time_.get_clock_type() != time.get_clock_type())
+    {
+      do_publish = true;
+    }
+    else if (!tf_publish_period_ || (tf_last_publish_time_ + *tf_publish_period_ < time))
```
Comment (fmauch) [inline pose_broadcaster/src/pose_broadcaster.cpp:159]: If we use ` < time`, the tf would only be published every second update if the controller rate is exactly the same as the tf publish rate, right?
```suggestion
    else if (!tf_publish_period_ || (tf_last_publish_time_ + *tf_publish_period_ <= time))
```

Comment (RobertWilbrandt) [reply pose_broadcaster/src/pose_broadcaster.cpp:159]: Good point, i didn't think of that

File: pose_broadcaster/doc/userdoc.rst [hunk 3]
```diff
@@ -0,0 +1,27 @@
+:github_url: https://github.com/ros-controls/ros2_controllers/blob/{REPOS_FILE_BRANCH}/pose_broadcaster/doc/userdoc.rst
+
+.. _pose_broadcaster_userdoc:
+
+Pose Broadcaster
+--------------------------------
+Broadcaster for poses measured by a robot or a sensor.
+The published message type is ``geometry_msgs/msg/PoseStamped``.
```
[the remark this answers is not in the document]
Comment (RobertWilbrandt) [reply pose_broadcaster/doc/userdoc.rst]: Good point, i updated this and the controller plugin file to mention that.

**Review by fmauch — APPROVED** [review]
Looks good to me! As @RobertWilbrandt wrote, we tested this in the UR driver already.

**Comment by bmagyar** [comment]
Would you see use of this being a chainable controller at the same time?

**Comment by RobertWilbrandt** [comment]
I don't have much experience with chaining controllers, but this sounds like a good extension. I can work on that and extend this PR.

**Comment by fmauch** [comment]
@bmagyar This should probably be backported? The semantic component has already been backported.

**Comment by christophfroehlich** [comment]
@mergifyio backport humble iron

**Pull request merged**